### PR TITLE
Entity Card: Card Addition

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

Entity Card - Similar to Sensor Card, except you do not have to use a Sensor Entity. This PR is similar to many of the current PRs. We are moving away from having to use extra sensors to get Attribute Data

The card will allow users to display state data or attribute data visually in a card

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
- cards:
   - attribute: brightness
      default: '0'
      entity: light.bedroom
      type: entity
      unit: '%'
   - attribute: current_temperature
      entity: climate.ecobee
      type: entity
      unit: °F
   - entity: sensor.processor_use
      name: Processor
      type: entity
  type: horizontal-stack
- type: entity
   entity: weather.home
- type: entity
   entity: cover.kitchen_window
- cards:
    - type: entity
       entity: device_tracker.demo_paulus
    - type: entity
       entity: device_tracker.demo_anne_therese
  type: horizontal-stack
```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: translations/en.json
Comment (zsarnett) [inline translations/en.json]: I am not sure I did Translations correctly. I can't seem to get them to show in the frontend. Just blanks..

Comment (balloob) [reply translations/en.json]: Change `src/translations/en.json`

Comment (zsarnett) [reply translations/en.json]: smh changed the wrong translations...

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: The downside of this approach imo is that we put a lot of work on the users, instead of us doing the right thing. This method is a step in the right direction, but it's still not aggressive enough. We can do better.

Let's start with creating pre-made configurations.

You pick a climate entity: offer the user a dropdown if they want to see the HVAC Mode, action, current temperature, current humidity (if available) or target temp.

Then there is also a custom mode where it's all whatever. Let's start with just the custom mode in this PR: give us an attribute or we just render the state. No data processing there. If it's state we're rendering, we can default to unit to the unit_of_measurement attribute, otherwise empty.

File: src/common/entity/compute_state_value.ts
Comment (balloob) [inline src/common/entity/compute_state_value.ts]: Did you copy paste this function from somewhere?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: So you are saying to remove any processing to make the data look nice. Ex. `off` -> `Off`
And remove the ability to give a Unit of measurement

File: src/common/entity/compute_state_value.ts
Comment (zsarnett) [reply src/common/entity/compute_state_value.ts]: Uh yes. 🙂. State Display function appended the UOM which I didn't want

Comment (balloob) [reply src/common/entity/compute_state_value.ts]: State and unit are connected. You can't just do one and have another function replicating the same logic just to see what the unit is.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Remove all the magic.

File: src/common/entity/compute_state_value.ts
Comment (zsarnett) [reply src/common/entity/compute_state_value.ts]: They are connected, sure. But the ability to change that UOM is not foreign

**Comment by bramkragten** [comment]
This is almost if not completely the same as a sensor card without a graph? Why not combine the 2?

**Comment by bramkragten** [comment]
We have `prefix` and `suffix` in other cards should we align that? Or...?

**Comment by balloob** [comment]
Please put this on hold until we have shipped media player. One beast at a time.

**Comment by zsarnett** [comment]
@balloob Where are we going with this :) I am still unsure. Maybe some time we talk this through? It sounds like a cool idea you have but not sure the full scope.

**Comment by balloob** [comment]
Let's jump on a call tomorrow. Maybe include Bram too.

**Comment by zsarnett** [comment]
Sounds good. Let me know a time when you are available. I have no other meetings today

**Comment by bramkragten** [comment]
Can you guys make it in 45 min?

**Comment by zsarnett** [comment]
@bramkragten I can make it. 12 PM EST. @balloob 9:00 AM for you?

**Comment by zsarnett** [comment]
So from our conversation this morning:

- I added the ability to configure a unit. To make this a fully manual card as discussed
- Added the ability to set a header and footer
- No magic is involved

Future PRs will include

- Making graph an option for the footer/header
- Converting the sensor card to an entity card with a footer

**Comment by zsarnett** [comment]
Should Default be an option here? We talked about this a bit in discord but what should we do here if, for example, the attribute doesn't exist if the state is off

Should we fall back to the State?

Should we render something that says no data?

With this current version of this card is all manual. Maybe we do have a default option. In the future we can set defaults automatically. Example: Hvac humidity - If the hvac is off and it doesn't exists we render Off.

**Comment by balloob** [comment]
If it's all manual, we should not do magic.
 - If state is unknown or unavailable, we can just render that.
 - If attribute doesn't exist, treat it as unknown. Do not render unit.
 - Otherwise render as usual.

**Comment by zsarnett** [comment]
So if the attribute doesn't exist. Render the value as "Unknown" (Localized)?

**Comment by zsarnett** [comment]
I also still feel like we should localize the state. I wouldn't consider Localizing into a  friendly state name magic. Aka (not_home -> Not Home) also this removes all other languages. So now they see the English version of the state since that's how we developed it

**Comment by zsarnett** [comment]
@balloob I left a few comments above about the recent changes. Would you be able to review sometime?

**Comment by zsarnett** [comment]
Also, What should the documentation look like? Not sure how this card should be marketed anymore

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Little easier to read if you use a ternary statement.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: This check means that now if you want to show no unit, you need to put in a space.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Why this check?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Now we're throwing away and re-creating this element at every render. And since `hass` is a property, that's a lot. You should create this element in `setConfig` and update it in `updated` with new `hass`

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Why are both column ?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Isn't this going to mess with all the colors that people pick ?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: paper-item ?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Let's not use IDs, just classes.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: You only need to use line-clamp if you want multi-line overflow ellipsis effect. Since this is 1 line, you can remove this.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Not a fan of extracting this. It makes reading the CSS a bit more complicated as it now has a dependency on the right classes being set in the HTML.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Why is it position relative ?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: And since it's inside `.name`, which also has an opacity, does it mean that `.name > span` is shown with `0.64` opacity?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Why this?

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Seems like I need to go through a rewrite the CSS for this card. This was copied from the sensor card so most of these things you are reviewing I don't know either :)

File: src/translations/en.json
Comment (balloob) [inline src/translations/en.json]: This one already exists 4 lines above it?

Comment (balloob) [reply src/translations/en.json]: I also don't see it being used?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: We're mixing rem and em.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Why do this in em? Are you trying to make this card grow proportionally when the user increases font size?

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: I didn't make this decision 🙂

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: What would you suggest here?

```ts
${this._config.unit || this._config.attribute ? "" : stateObj.attributes.unit_of_measurement}
```

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: A config variable for show unit?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: This is straight from the entities card. Should we not check this here?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: We should probably update the entities card then, no?

Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Nope, the entities card has a setter for `hass`. It only re-renders when new config arrives, otherwise it manually looks at the DOM elements and finds the ones that need `hass` updated.

Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: You cannot do that here because you are rendering a template based on hass-data. Entities card just needs to pass it along to the rows/header/footer

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Probably but this is how the sensor card has been since creation. I like the greyed look of the header. Gives it depth and separation from the value

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Not sure what the Variable should be here?

Seems this is what is used

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Got it. Makes sense

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: I removed the span in name as it was unneeded

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: @bramkragten what should we use?

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: That's why we have secondary text color as a variable.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-entity-card.ts]: `--state-icon-color`

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: Make sure you unset it, or else removing a footer in a config update does nothing.

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]:
```suggestion
    return 2 + (this._config.footer ? 1 : 0);
```

Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Or maybe ask the footer for size ? (although I don't think we implemented that)

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: This will filter out if the value is falsey, like 0. 
```suggestion
      ? this._config.attribute in stateObj.attributes
```

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [inline src/panels/lovelace/cards/hui-entity-card.ts]: handle attribute being `0`

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (balloob) [reply src/panels/lovelace/cards/hui-entity-card.ts]: `"unit" in this._config`

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: Doesnt the above do the same? This would just be an extra step?

```ts
"unit" in this._config ? this._config.unit : this._config.attribute ? "" : stateObj.attributes.unit_of_measurement
```

because if the attribute is set then we don't want the state measurement

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: I guess maybe 

```ts
!("unit" in this._config) && this._config.attribute ? "" : this._config.unit || stateObj.attributes.unit_of_measurement
```

File: src/panels/lovelace/cards/hui-entity-card.ts
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-entity-card.ts]: This is probably something we should implement in the future as the graph footer is much bigger than the buttons. But for now.

**Review by balloob — APPROVED** [review]

**Comment by balloob** [comment]
Ok to merge when docs added !

**Pull request merged**
